File: 2023/FinalBuild/src/main/cpp/Pigeon.cpp

#include "Pigeon.h"
#include <frc/smartdashboard/SmartDashboard.h>
#include <frc/geometry/CoordinateAxis.h>
#include <units/velocity.h>
#include <units/angle.h>
#include <frc/geometry/Translation2d.h>
#include <ctre/phoenix/platform/Platform.hpp>

using namespace ctre::phoenix6::hardware;

Pigeon::Pigeon()
{

  RobotPigeon.SetFusedHeading(0.0);
  RollOffset = -units::degree_t{RobotPigeon.GetRoll()};

}

Pigeon& Pigeon::GetInstance()
{
  static Pigeon Instance;
  return Instance;
}

PigeonData Pigeon::GetPigeonData()
{

  PigeonIMU::FusionStatus fusionStatus;
  double xyz_degrees_per_s[3];
  RobotPigeon.GetRawGyro(xyz_degrees_per_s);
  RobotPigeon.GetFusedHeading(fusionStatus);
  units::degree_t fusionDegrees{fusionStatus.heading};

  return PigeonData( fusionDegrees, 
                     xyz_degrees_per_s[2],
                     (RobotPigeon.GetState() == PigeonIMU::PigeonState::Ready));
}

units::degrees_per_second_t Pigeon::GetYawRate()
{
  double xyz[3];
  RobotPigeon.GetRawGyro(xyz);
  return units::degrees_per_second_t{xyz[2]};
}

units::degree_t Pigeon::GetRoll()
{
  static double alpha = 0.75;

  RollFiltered = (RobotPigeon.GetRoll() * alpha) + (RollFiltered * (1 - alpha));

  return units::degree_t{RollFiltered};
}

units::degree_t Pigeon::GetPitch()
{
  return units::degree_t{RobotPigeon.GetPitch()};
}

units::degree_t Pigeon::GetCorrectedRoll()
{
  return Pigeon::GetRoll() + RollOffset;
}

void Pigeon::RestPigeonAngle()
{
  RobotPigeon.SetFusedHeading(0.0);
}

File: 2020/Team4005_2020/src/main/java/frc/robot/subsystems/Pigeon.java
package frc.robot.subsystems;

import com.ctre.phoenix.sensors.PigeonIMU;
import com.ctre.phoenix.motorcontrol.can.TalonSRX;

import frc.robot.RobotConstants;

public class Pigeon
{
  private static final TalonSRX pigeonTalon = new TalonSRX(RobotConstants.PIGEON_TALONID);
  public static final Pigeon instance = new Pigeon(pigeonTalon);
  private static PigeonIMU pigeon;

  Pigeon(TalonSRX talon)
  {
    pigeon = new PigeonIMU(talon);
    pigeon.setFusedHeading(0.0, 50);
  }

  public PigeonData getPigeonData()
  {
    PigeonIMU.GeneralStatus genStatus = new PigeonIMU.GeneralStatus();
    PigeonIMU.FusionStatus fusionStatus = new PigeonIMU.FusionStatus();
    double [] xyz_dps = new double[3];

    pigeon.getGeneralStatus(genStatus);
    pigeon.getRawGyro(xyz_dps);
    pigeon.getFusedHeading(fusionStatus);

    return new PigeonData(fusionStatus.heading, xyz_dps[2], (pigeon.getState() == PigeonIMU.PigeonState.Ready));
  }
}